### DISCUSSION TITLE
Unable to upgrade from 3.1.6 to 3.2.3

### DISCUSSION THREAD

**Discussion by svetlinski** [netbox-community/netbox#9428] [Q&A]
### NetBox version

v3.2.3

### Python version

3.8

### Steps to Reproduce

Apologies if my post is not in the correct section of the site.

I followed all the steps to upgrade from 3.1.6 to 3.2.3 and I believe I haven't missed anything. I kept on getting various errors and I had to install additional Django modules, however, now, I seem to be getting the attached error. For some reason, it looks like I am now using NginX instead of Apache, but that is not much of an issue. Yes, I have already updated the files in sites-available. 
![image](https://user-images.githubusercontent.com/99681269/170381131-817ce32b-ceec-4e9c-a68f-d020616dd159.png)
If I stop NginX and start Apache, I just get a "x.x.x.x refused to connect"

I have already ensured that I am running PSQL version 12.11 and I have manually ran the migration about 5 times in total.
From what I understand. there seems to be a column missing in the database - object_type? Where would that have gone? I have not deleted it manually...

netbox=> SELECT * FROM extras_customfield;
 id | type | name | label | description | required | filter_logic | default | weight | validation_minimum | validation_maximum | validation_regex | choices | created | last_updated
----+------+------+-------+-------------+----------+--------------+---------+--------+--------------------+--------------------+------------------+---------+---------+--------------
(0 rows)

(END)

Any help is much appreciated.

### Expected Behavior

Login page should load

### Observed Behavior

Error page loads instead.

**Comment by candlerb** [comment]
> I kept on getting various errors and I had to install additional Django modules

No. No, no, no.  If you get errors, you should *not* install any additional Django modules; if you do, it means you're running the upgrade wrongly.  The upgrade builds a new virtualenv with all required modules.  Your workarounds are just compounding the problem.

As for `extras_customfield`, these are the columns it should have:

```
netbox3=# select * from extras_customfield limit 0;
 id | type | name | label | description | required | weight | filter_logic | choices | default | validation_maximum | validation_minimum | validation_regex | created | last_updated | object_type_id
----+------+------+-------+-------------+----------+--------+--------------+---------+---------+--------------------+--------------------+------------------+---------+--------------+----------------
(0 rows)
```

This means that some previous migration has failed.

However, the Netbox 3.2 codebase has no migrations relating to extras_customfield at all:

```
netbox $ grep -R extras_customfield .
netbox $
```

This is because all migrations since v2.11 were deleted in v3.0; migrations of older versions of Netbox must go v2.x -> v2.11 -> v3.x, as [documented](https://github.com/netbox-community/netbox/releases/tag/v3.0.0).

What I can say for sure is that your system is in a terribly broken state.

If I had to guess at what has gone wrong, I would say that maybe you've tried upgrading Netbox by untarring or unzipping a new release on top of an old one. *This does not work* because it does not remove old files which should not be there.  That's the only way I can think you've got any migrations which refer to `extras_customfield`, whilst you claim this is Netbox 3.2.

There are [two documented ways](https://docs.netbox.dev/en/stable/installation/upgrading/) to perform an upgrade: one is to to untar a new release into a *completely fresh* directory and to change a symlink to point to it; the other is to do a `git pull` on a checkout of the master branch.  Which one applies to you depends on which way you originally installed Netbox.

I am also guessing that your original Netbox install was an older one (say 2.9) and then you upgraded to 3.x, and maybe you skipped the intermediate step of migrating to 2.11, and ignored all the errors which would have been thrown up.

As I don't know the history of your system, this is really just guesswork on my part.  However if you really have gone through at least two broken upgrades, the only thing I can really recommend at this point is to install Netbox 3.2.3 from scratch, and to repopulate it with data from scratch.  Then going forward, make sure you follow the upgrade documentation, and if you see any error, *stop there* and solve the problem properly, rather than trying to patch around it.

**Reply by svetlinski** [reply]
Many thanks for your response and advice. 

I can assure you that we have only 1 previous version and that is 3.1.6. It is still on the server. 
This is the reason why my colleagues cloned the virtual machine as well as created a snapshot of it. I will look into restoring a backup as we had over 20 sites populated in there and there is no way we can start to populate it from scratch. 

Once restored, I will have to try the upgrade again - I think you may be right that the untar had gone wrong, although I can see the separate folder on the server where I extracted the new release, so not entirely sure where things went wrong. 

I will update you again shortly, once I restore the backup and try the upgrade again.

Thank you and have a great weekend.

**Reply by candlerb** [reply]
OK, hope it works out (reverting to a known good backup is always a good idea).

BTW I was wrong about customfield migrations in v3.2.3, they do exist - I was just searching for the wrong term.

```
$ grep -R customfield netbox/*/migrations
netbox/dcim/migrations/0146_modules.py:        ('extras', '0066_customfield_name_validation'),
netbox/extras/migrations/0072_created_datetimefield.py:            model_name='customfield',
netbox/extras/migrations/0069_custom_object_field.py:            model_name='customfield',
netbox/extras/migrations/0061_extras_change_logging.py:            model_name='customfield',
netbox/extras/migrations/0061_extras_change_logging.py:            model_name='customfield',
netbox/extras/migrations/0066_customfield_name_validation.py:            model_name='customfield',
netbox/extras/migrations/0002_squashed_0059.py:        ('extras', '0010_customfield_filter_logic'),
netbox/extras/migrations/0002_squashed_0059.py:        ('extras', '0029_3569_customfield_fields'),
netbox/extras/migrations/0002_squashed_0059.py:        ('extras', '0042_customfield_manager'),
netbox/extras/migrations/0002_squashed_0059.py:        ('extras', '0050_customfield_changes'),
netbox/extras/migrations/0002_squashed_0059.py:        ('extras', '0051_migrate_customfields'),
netbox/extras/migrations/0002_squashed_0059.py:        ('extras', '0052_customfield_cleanup'),
netbox/extras/migrations/0067_customfield_min_max_values.py:        ('extras', '0066_customfield_name_validation'),
netbox/extras/migrations/0067_customfield_min_max_values.py:            model_name='customfield',
netbox/extras/migrations/0067_customfield_min_max_values.py:            model_name='customfield',
netbox/extras/migrations/0068_configcontext_cluster_types.py:        ('extras', '0067_customfield_min_max_values'),
netbox/extras/migrations/0071_standardize_id_fields.py:            model_name='customfield',
netbox/extras/migrations/0001_squashed.py:import extras.models.customfields
netbox/extras/migrations/0001_squashed.py:                ('objects', extras.models.customfields.CustomFieldManager()),
```

I couldn't find where the object_type_id field is created.

I'd be interested to see the full backtrace from your failing Netbox instance (set EMAIL and ADMINS in configuration.py, and it will be mailed out to you; or set DEBUG=True)

But I'm wondering now, if you found you had to add extra python packages (and therefore you weren't running inside Netbox's virtualenv), maybe you were running the wrong version of Django at the time the migrations ran, and therefore the migrations didn't work as expected.

**Reply by svetlinski** [reply]
You may be right about the packages, however, I had to add those outside the virtualenv, as well as within. Unfortunately, we already performed the restore and probably wiped all the logs in the process. 

I will re-attempt the upgrade process soon and will let you know how I get on. 
I remember that during installation, I noticed Netbox is running inside /opt/netbox/netbox-x.x.x/netbox/netbox/ which seemed a little odd to me, but it works, so I didn't touch it further.
When extracting the new version, I ensured it is untared in a completely separate directory, rather than overwrite the existing one, so this is probably what has caused the break. I had to go into a lot of .conf files and change the default paths, for example, the nginx file, where there is an alias for /static. It was pointed to /opt/netbox/xyz when in fact, in my case, it would need to point to /opt/netbox/netbox-x.x.x/netbox/netbox/static (I might have one too many "netbox" folders in there, but you get the gist).

**Reply by candlerb** [reply]
> You may be right about the packages, however, I had to add those outside the virtualenv, as well as within.

That should never be required, and means there is an underlying problem.  Possibly your startup script (e.g. systemd unit file) is pointing to the wrong place. It should be:

```
ExecStart=/opt/netbox/venv/bin/gunicorn --pid /var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config /opt/netbox/gunicorn.py netbox.wsgi
```

Basically, Netbox should *never* use any globally-installed packages, only those in its venv.

> I will re-attempt the upgrade process soon and will let you know how I get on. I remember that during installation, I noticed Netbox is running inside /opt/netbox/netbox-x.x.x/netbox/netbox/ which seemed a little odd to me

If you follow the standard install instructions, you will have `/opt/netbox-x.x.x`, and a symlink from `/opt/netbox` to that. The virtualenv will be created as `/opt/netbox-x.x.x/venv`, but also accessible as `/opt/netbox/venv` thanks to the symlink.

Note that there are nested directories called "netbox", and in particular the config file is `/opt/netbox/netbox/netbox/configuration.py` when in its standard location.
